sys/event/event.c

```diff
@@ -60,29 +88,79 @@ void event_cancel(event_queue_t *queue, event_t *event)
     irq_restore(state);
 }
 
-event_t *event_get(event_queue_t *queue)
+void event_counted_cancel(event_queue_t *queue, event_counted_t *event,
+                          bool cancel_all)
 {
+    assert(queue);
+    assert(event);
+
     unsigned state = irq_disable();
-    event_t *result = (event_t *) clist_lpop(&queue->event_list);
+    if (cancel_all) {
+        event->num_posted = 1;
+    }
+
+    if (event->num_posted && --event->num_posted == 0) {
+        clist_remove(&queue->event_list, &event->list_node);
+    }
     irq_restore(state);
+}
+
+static event_t *_get(event_queue_t *queue)
+{
+    event_t *result = (event_t *)clist_lpop(&queue->event_list);
 
     if (result) {
         result->list_node.next = NULL;
     }
+
     return result;
 }
 
-event_t *event_wait_multi(event_queue_t *queues, size_t n_queues)
+static event_counted_t *_get_counted(event_queue_t *queue)
+{
+    clist_node_t *node = clist_lpeek(&queue->event_list);
+    event_counted_t *result = container_of(node, event_counted_t, list_node);
+
+    if (result) {
+        assert(result->num_posted);
+        if (--result->num_posted == 0) {
+            clist_lpop(&queue->event_list);
```

Comment:
```suggestion
            clist_lpop(&queue->event_list);
            result->list_node.next = NULL;
```

Reply:
Not really necessary, but I suppose it doesn't hurt

Reply:
This is done for the non-counting event. IMO, we should have this consistent. (And maybe there is code that checks if an event is queued by comparing the `next` pointer with `NULL`.)